### ISSUE TITLE
Can't use root mode in versions 3.x.x

### ISSUE THREAD

**Issue by Zafrkant** [MuntashirAkon/AppManager#948]
### Please check before submitting an issue

- [X] I know what my device, OS and App Manager versions are
- [X] I know how to take logs
- [X] I know how to reproduce the issue which may not be specific to my device

### Describe the bug

I'm using Xiaomi Mi 11i, Xiaomi EU 13.0.13.0, App Manager Debug 2130.
I've tried multiple 3.x.x versions but results are the same, I can't use root mode. On v2.6.5.1 everything is working fine.
Full system log will be provided.

### To Reproduce

- 1. Install debug version
- 2. Select root mode.
- 3. Settings | Mode of operation | Root

### Expected behavior

No root access will be provided.

### Logs

Log is provided via Mega due to it's large size, GitHub: There was an error creating your Issue: body is too long, body is too long (maximum is 65536 characters).
https://mega.nz/file/e5clkSSB#hvEJ_JSzxOQnmcPzXHC5uKnEfK8VOe4n3bifOGxyZL4

### Device info

- Device: Xiaomi Mi 11i
- OS Version: Xiaomi EU 13.0.13.0
- App Manager Version: AM Debug v3.1.0-alpha02 Run#2130
- Mode: Root/ADB/NonRoot: Root

**Comment by MuntashirAkon** [comment]
```
617:696 E/vold: Failed to set project id on /data/media/0/Android/data/io.github.muntashirakon.AppManager.debug/: Operation not supported on transport endpoint
30553:30553 W/ContextImpl: Failed to ensure /storage/emulated/0/Android/data/io.github.muntashirakon.AppManager.debug/cache: android.os.ServiceSpecificException:  (code -1)
30745:30745 D/AndroidRuntime: >>>>>> START com.android.internal.os.RuntimeInit uid 0 <<<<<<
30553:30689 E/RootService: main.jar: open failed: EROFS (Read-only file system)
java.io.FileNotFoundException: main.jar: open failed: EROFS (Read-only file system)
	at libcore.io.IoBridge.open(IoBridge.java:575)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:236)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:186)
	at io.github.muntashirakon.AppManager.ipc.RootServiceManager.lambda$startRootProcess$1(RootServiceManager.java:193)
	at io.github.muntashirakon.AppManager.ipc.RootServiceManager$$ExternalSyntheticLambda3.run(Unknown Source:9)
	at io.github.muntashirakon.AppManager.ipc.RootService.lambda$asRunnable$0(RootService.java:201)
	at io.github.muntashirakon.AppManager.ipc.RootService$$ExternalSyntheticLambda0.run(Unknown Source:2)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:920)
Caused by: android.system.ErrnoException: open failed: EROFS (Read-only file system)
	at libcore.io.Linux.open(Native Method)
	at libcore.io.ForwardingOs.open(ForwardingOs.java:567)
	at libcore.io.BlockGuardOs.open(BlockGuardOs.java:273)
	at libcore.io.ForwardingOs.open(ForwardingOs.java:567)
	at android.app.ActivityThread$AndroidOs.open(Unknown Source:15)
	at libcore.io.IoBridge.open(IoBridge.java:561)
	... 9 more
```

It is failing because the internal storage in your device is set to read-only for App Manager which shouldn't happen. Check if any of the other software is messing with App Manager, examples include Xposed or similar frameworks and Storage Redirect. Also check if the path `/storage/emulated/0/Android/data/io.github.muntashirakon.AppManager.debug/` exists.

**Comment by MuntashirAkon** [comment]
> Do you have idea how can I check what is making storage read-only?

Not sure. You could do onething: Run AM Debug in no-root mode, grant file permissions. Then open the app details page of AM Debug, navigate to Activities tab, find FmActivity and open it to see if it can display anything.

**Comment by Zafrkant** [comment]
Thank you for answer.
I've did as you suggested and on click it is opening file manager on /storage/emulated/0 path.
![signal-2022-12-05-135343](https://user-images.githubusercontent.com/30977345/205659357-8f2d25e2-050e-4f8d-94a6-1c0def19b3e5.jpeg)

**Comment by unRekable** [comment]
I have the same issue. I'm also using a Xiaomi device (Mi 11 Ultra).
The error I have is the same: open failed erofs (read-only file system)

As path, I'm having the same as @Zafrkant when opening FmActivity.

**Comment by unRekable** [comment]
By the way, in case it matters, when opening FmActivity, I don't have permissions to view /android/data - probably that's causing the problem?

**Comment by unRekable** [comment]
I've updated to Android 13 and MIUI 14 today - still the same issue. Any update on this @MuntashirAkon ?

**Comment by Zafrkant** [comment]
I've upgraded Xiaomi EU to v14.0.3.0, App Manager to v3.0.4 and it is working fine now.

**Comment by unRekable** [comment]
Hope it keeps working for you. Still doesn't work for me on MIUI 14 and Android 13. I will install the latest weekly and see.
On Miui 13 and Android 12 one time it showed me root mode and asked me to select a storage for backups or something, don't entirely remember. Though, I think I disallowed root in first, restarted and allowed it then. Whatever, it just showed me the root mode for a short time, when swichting modes it didn't work anymore.

**Comment by MuntashirAkon** [comment]
> Any update on this @MuntashirAkon ?

Sorry, no updates. If you're willing to help, I will publish some debug builds to figure out the issue first.

**Comment by unRekable** [comment]
> > Any update on this @MuntashirAkon ?
> 
> Sorry, no updates. If you're willing to help, I will publish some debug builds to figure out the issue first.

Sure, would appreciate to help figuring this out.

**Comment by unRekable** [comment]
Here's the log file/export - in case it helps you. Installed the latest build today. 

https://file.io/0wcoXlniJ95j

**Comment by unRekable** [comment]
Can you build a debug build @MuntashirAkon ? Been waiting long 😄

**Comment by Bladius2022** [comment]
![Screenshot_20230224-064416_Solid Explorer](https://user-images.githubusercontent.com/112637245/221101650-c61ca250-b6fb-41a2-a2c0-bab4f69773d6.png)

**Comment by MuntashirAkon** [comment]
Try 35d8c3b0aa0b8604a604fc39730c436a256dbd53

**Comment by Bladius2022** [comment]
I don't understand what to check.  There is nothing to download from the link provided

**Comment by unRekable** [comment]
> I don't understand what to check. There is nothing to download from the link provided

Same. I never compiled into an apk file. If you can make a beta/test build, would be awesome @MuntashirAkon. Thanks!

**Comment by MuntashirAkon** [comment]
Read the docs: https://muntashirakon.github.io/AppManager/en/#subsec:binary-distribution-sources

**Comment by unRekable** [comment]
Downloaded the Debug Build from your telegram.

It crashes on start and cannot run. Here's the error log I could get:
https://pastebin.com/FqJYKXqY

Thanks!

**Comment by MuntashirAkon** [comment]
Try#2: 77b1f5dd1889b3c9126dd9b69428fbad9982c28b

**Comment by unRekable** [comment]
Same. Re-installed. It tells me the key, then ask for root and after I give permission it crashes / closes the app.

**Comment by Bladius2022** [comment]
> Downloaded the Debug Build from your telegram.
> 
> It crashes on start and cannot run. Here's the error log I could get: https://pastebin.com/FqJYKXqY
> 
> Thanks!

Where is link to telegram?

**Comment by MuntashirAkon** [comment]
> Same. Re-installed. It tells me the key, then ask for root and after I give permission it crashes / closes the app.

This is very weird. The shared storage clearly exists as per your previous comments, but App Manager is unable to use them somehow.

In a Terminal, run the following with root:

```
mount | grep Android/data
```

**Comment by unRekable** [comment]
![Uploading Screenshot_2023-02-28-18-59-17-564_com.termux.jpg…]()

**Comment by MuntashirAkon** [comment]
@unRekable: Screenshot was not uploaded properly.

**Comment by unRekable** [comment]
https://ibb.co/VjwwtP3

**Comment by MuntashirAkon** [comment]
Ah, it's a device mapper no. 33 (disk encryption?). Run the following command to check the mapping:

```
ls -l /dev/mapper/
```

Also, check if `/dev/media` is properly mounted:

```
mount | grep /dev/media
```

Run everything using root.

**Comment by unRekable** [comment]
/dev/mapper/ no such directory, hence it isn't mounted.

I think my device is encrypted yes. I always enter my pin to decrypt it in TWRP.

**Comment by MuntashirAkon** [comment]
> /dev/mapper/ no such directory, hence it isn't mounted.

And the output of the other command?

**Comment by MuntashirAkon** [comment]
> I think my device is encrypted yes. I always enter my pin to decrypt it in TWRP.

Use of dm-X usually means that the partitions are encrypted in this context. It's absolutely okay to /dev/mapper not containing anything as a result.

**Comment by unRekable** [comment]
> > /dev/mapper/ no such directory, hence it isn't mounted.
> 
> And the output of the other command?

The other command gave me no output
It couldn't grep something

**Comment by unRekable** [comment]
https://ibb.co/VD43gpF

It's 1| I think (the output)

**Comment by MuntashirAkon** [comment]
Being relatively new in Android, I'm not particularly familiar with how storage works in Android. But `/data/media` should, at least, exist and I'm not sure why this isn't the case for you Android 13. Maybe @mirfatif can help regarding this.

**Comment by unRekable** [comment]
/data/media folder exists. Just /dev/media and /dev/mapper doesn't exist

**Comment by unRekable** [comment]
"mount | grep /data/media" also returns me 1|

**Comment by mirfatif** [comment]
@MuntashirAkon the provided info doesn't seem to be very helpful.

`E/vold: Failed to set project id on ...` seems to be an issue at the kernel / filesystem level. `vold` sets this to manage quotas.

A fresh install Clearing 

`W/ContextImpl: Failed to ensure ...` indicates that you are using `getExternalCacheDir()` to open the file in the cache directory when `E/RootService: main.jar: open failed: EROFS (Read-only file system)` is thrown? Is it required to use the external cache directory? Did you try to use the files directory? Or maybe as a fallback use SAF to create a new directory in the root of the external storage?

But apparently the problem doesn't seem specifically with your app but the operating system b/c the standard APIs are broken. Or a framework mod is making them break.

**Comment by MuntashirAkon** [comment]
I didn't pay much attention to the `vold` error which is emitted from here: https://cs.android.com/android/platform/superproject/+/master:system/vold/Utils.cpp;l=245;drc=b0f997deae8edd0a3f7e3ebfc8da5b62bbea10c1

But it seems it could be an issue with the ROM itself as per this comment: https://github.com/remote-android/redroid-doc/issues/42#issuecomment-1211496773

**Comment by MuntashirAkon** [comment]
All issues seem to be coming from Xioami phones with A13 (Yet another Motorola-like situation (#880)?)

**Comment by mirfatif** [comment]
Xiaomi has a history of [interesting bugs](https://github.com/topjohnwu/Magisk/issues/1880#issuecomment-546657588).

**Comment by Bladius2022** [comment]
On RC1 app with root, crash on start

**Comment by MuntashirAkon** [comment]
> On RC1 app with root, crash on start

This is expected. It's incredibly difficult for me to support a potentially broken ROM.

**Comment by Bladius2022** [comment]
the ROM is not corrupted.  Since the problem is on the original Xiaomi ROM and Xiapmi.eu as well as on the ArrowOS 13 ROM, it is not a rom issue.

**Comment by MuntashirAkon** [comment]
> the ROM is not corrupted. Since the problem is on the original Xiaomi ROM and Xiapmi.eu as well as on the ArrowOS 13 ROM, it is not a rom issue.

Read the discussions above. `vold` (part of Android system) being unable to assign a project ID (permissions) for App Manager's cache directory (which, presumably, occurred due to `/data/media/0` being non-existent) is clearly a sign of broken system. This isn't necessarily the ROM itself though, since we have plenty of MIUI 13 and 14 users who are using App Manager without any issues. This is likely something else. We did find a few interesting issues in the past. For example, Lineage OS 17.1 users were unable to uninstall App Manager from their device as the action itself was causing the Android installer to crash. We initially thought this to be an issue with LOS itself but later only to find out that it was an issue with external SD Card and a certain OpenGapps version. Some Samsung users still cannot uninstall App Manager because One UI's installer crashes when you click uninstall. But in this case, One UI itself is to be blamed.

**Comment by Bladius2022** [comment]
puzzling.  Is it possible that the Permissions system app that is reinstalled by default could be buggy in a particular version?  MIUI 13 is only for Android 12 and MIUI 14 is for both Android 12 and Android 13. On MIUI 14 with A12 AppManager works for me, but does it also work on MIUI 14 with Android 13?

**Comment by MuntashirAkon** [comment]
> does it also work on MIUI 14 with Android 13?

It does, yes.

**Comment by unRekable** [comment]
> puzzling. Is it possible that the Permissions system app that is reinstalled by default could be buggy in a particular version? MIUI 13 is only for Android 12 and MIUI 14 is for both Android 12 and Android 13. On MIUI 14 with A12 AppManager works for me, but does it also work on MIUI 14 with Android 13?

Here it doesn't on A13 with MIUI 14 (Xiaomi.EU weekly version). I also have no idea what's causing it but my first guess was it cannot access the /android/media folder. Atleast it is empty when opening it with FmActivity.

I also don't use much on my phone, basically Magisk alpha (zygisk), AccA (for battery management), LSPosed (zygisk),  Systemless Hosts from Magisk, and some stuff I installed (after the App Manager).

**Comment by MuntashirAkon** [comment]
> I also have no idea what's causing it but my first guess was it cannot access the /android/media folder. Atleast it is empty when opening it with FmActivity.

It's the `/data/media` folder, accessible only via root for non-system users. This folder contains the shared storages, SD Card, etc. from all users so that the system can access all of the available storages without worrying about any potential privacy leaks. MIUI 14 has already received February updates in many devices, I think. If updating to the latest version does not work for you, you should consider disabling Magisk temporarily to sort out the issue (booting in safe mode, perhaps?). If it's not an issue with the ROM, then disabling Magisk should fix the issues.

**Comment by Bladius2022** [comment]
![Screenshot_20230308-101047_Root Browser](https://user-images.githubusercontent.com/112637245/223671321-51c53bfa-8ffa-471e-88fe-5b6eeddc5e2d.png)
DATA/MEDIA i have access

**Comment by kryptobolt200528** [comment]
> > puzzling. Is it possible that the Permissions system app that is reinstalled by default could be buggy in a particular version? MIUI 13 is only for Android 12 and MIUI 14 is for both Android 12 and Android 13. On MIUI 14 with A12 AppManager works for me, but does it also work on MIUI 14 with Android 13?
> 
> Here it doesn't on A13 with MIUI 14 (Xiaomi.EU weekly version). I also have no idea what's causing it but my first guess was it cannot access the /android/media folder. Atleast it is empty when opening it with FmActivity.
> 
> I also don't use much on my phone, basically Magisk alpha (zygisk), AccA (for battery management), LSPosed (zygisk), Systemless Hosts from Magisk, and some stuff I installed (after the App Manager).

This is unrelated to the issue but i would strongly would recommend you not to use any 3rd party magisk version especially magisk alpha the developer pf magisk alpha has an history of trying to suggest unacceptable behaviour of magisk such as deafult crash reports being sent to the servers.

**Comment by unRekable** [comment]
Newest release doesn't work
Here is the error log: https://pastebin.com/CFcfVRwR

**Comment by MuntashirAkon** [comment]
> Newest release doesn't work

Yes, it is deliberately done so. As stated earlier, try booting in safe mode without Magisk to see if it works. I cannot solve an issue if I do not know the reason. However, logs indicate that it's an issue with your OS or, more likely, due to some illegal modifications made by a third-party app.

**Comment by unRekable** [comment]
I can't run apps in safe mode. 

My current seup is Magisk with:
- ACC, MiXplorer, Magisk Systemless Hosts 
- ReVanced Manager (+Youtube) with Vanced G.

I had LSPosed installed before which I removed. 

I really have no idea what could cause it or how to debug/troubleshoot it.
If you have any ideas on troubleshooting it, please do let me know.

**Comment by unRekable** [comment]
I remember once the root mode was displayed in the settings.
I don't know what happened, I think I refused root rights when starting after the reinstallation, then was asked again directly (which I accepted) - then I was also asked for a backup path or something (which is not known to me). I accidentally clicked away too quickly and then saw in the settings that I use the root mode.
I wanted to verify this again, restarted the app and then the root mode was gone.

**Comment by MuntashirAkon** [comment]
> I really have no idea what could cause it or how to debug/troubleshoot it.
> If you have any ideas on troubleshooting it, please do let me know.

As I said earlier, I don't see a way to figure out the issue without unrooting the device.

I have received a crash report from another user (Xioami, too). Not sure if he's one of the three users who have already reported it.

<details>

<summary>Crash report</summary>

```
java.lang.IllegalStateException: External directory unavailable.
at io.github.muntashirakon.AppManager.ipc.RootServiceManager.startRootProcess(RootServiceManager.java:161)
at io.github.muntashirakon.AppManager.ipc.RootServiceManager.createBindTask(RootServiceManager.java:285)
at io.github.muntashirakon.AppManager.ipc.RootService.bindOrTask(RootService.java:146)
at io.github.muntashirakon.AppManager.ipc.RootService.bind(RootService.java:110)
at io.github.muntashirakon.AppManager.ipc.RootService.bind(RootService.java:125)
at io.github.muntashirakon.AppManager.ipc.ServiceConnectionWrapper.lambda$startDaemon$0$io-github-muntashirakon-AppManager-ipc-ServiceConnectionWrapper(ServiceConnectionWrapper.java:117)
at io.github.muntashirakon.AppManager.ipc.ServiceConnectionWrapper$$ExternalSyntheticLambda0.run(Unknown Source:4)
at android.os.Handler.handleCallback(Handler.java:942)
at android.os.Handler.dispatchMessage(Handler.java:99)
at android.os.Looper.loopOnce(Looper.java:201)
at android.os.Looper.loop(Looper.java:288)
at android.app.ActivityThread.main(ActivityThread.java:7872)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:946)
Caused by: java.io.FileNotFoundException: No available shared storage found.
at io.github.muntashirakon.AppManager.utils.FileUtils.getExternalCachePath(FileUtils.java:210)
at io.github.muntashirakon.AppManager.ipc.RootServiceManager.startRootProcess(RootServiceManager.java:159)
at io.github.muntashirakon.AppManager.ipc.RootServiceManager.createBindTask(RootServiceManager.java:285)
at io.github.muntashirakon.AppManager.ipc.RootService.bindOrTask(RootService.java:146)
at io.github.muntashirakon.AppManager.ipc.RootService.bind(RootService.java:110)
at io.github.muntashirakon.AppManager.ipc.RootService.bind(RootService.java:125)
at io.github.muntashirakon.AppManager.ipc.ServiceConnectionWrapper.lambda$startDaemon$0$io-github-muntashirakon-AppManager-ipc-ServiceConnectionWrapper(ServiceConnectionWrapper.java:117)
at io.github.muntashirakon.AppManager.ipc.ServiceConnectionWrapper$$ExternalSyntheticLambda0.run(Unknown Source:4)
at android.os.Handler.handleCallback(Handler.java:942)
at android.os.Handler.dispatchMessage(Handler.java:99)
at android.os.Looper.loopOnce(Looper.java:201)
at android.os.Looper.loop(Looper.java:288)
at android.app.ActivityThread.main(ActivityThread.java:7872)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:946)

Device Info:
App version: 3.1.0-rc02
App version code: 421
Android build version: 1677109697
Android release version: 13
Android SDK version: 33
Android build ID: TQ1A.230205.002 release-keys
Device brand: Redmi
Device manufacturer: Xiaomi
Device name: sweet
Device model: M2101K6G
Device product name: sweet
Device hardware name: qcom
ABIs: [arm64-v8a, armeabi-v7a, armeabi]
ABIs (32bit): [armeabi-v7a, armeabi]
ABIs (64bit): [arm64-v8a]
System language: ru-RU
In-App Language: auto
Mode: auto
Inferred Mode: root
```

</details>

**Comment by unRekable** [comment]
What can I do exactly to help you identify the issue? I'm able to run AM without root privilege granted.

**Comment by unRekable** [comment]
If it helps you, this weekend I'll update the xiaomi.eu Rom and can leave it unrooted for testing 

Just tell me what I have to do then

**Comment by MuntashirAkon** [comment]
> If it helps you, this weekend I'll update the xiaomi.eu Rom and can leave it unrooted for testing
> 
> Just tell me what I have to do then

In no-root mode, just launching the app should work as expected. There are many EU ROM testers who reported it to be working.

1. In the unrooted OS, install App Manager, check if it launches as expected.
2. If it worked as expected, root your device and try again:
   i. Without granting root to App Manager
   ii. With root.
3. If both worked, install/flash whatever you were flashing one by one and see if App Manager is reporting any issues.

This should pin-point exactly where it went wrong.

Note that there are always the workarounds, but we should use it as the last resort and find exactly what's causing the issue.

**Comment by unRekable** [comment]
Hey! Updated my room today.

So, 1. works, 2. i. also works, 2. ii. unfortunately doesn't work. 

After rooting the device, I disabled every module on Magisk (of the very few), also disabled Zygisk on Magisk and AM couldn't run.

I had LSPosed installed before but don't use it anymore since a while. You don't think any changes could have been left over?

I mean it's an "external directory unavailable" error, for me it still looks like it cannot access /android/data/ directory where everything else can access this folder (afaik). At least the folder is also empty in FmActivity.

Thanks!

**Comment by unRekable** [comment]
By the way, you said "In the unrooted OS, install App Manager, check if it launches as expected."
I did not re-install it after updating my rom, should not make a change, right?

**Comment by MuntashirAkon** [comment]
> 1. works, 2. i. also works, 2. ii. unfortunately doesn't work.

Thanks for testing! If 2 (i) works but 2 (ii) does not, it's likely an issue with App Manager itself. I'll see if I can find any issues. In the mean time, can you try launching the app with ADB? This might clear things up a bit.

**Comment by Bladius2022** [comment]
On ArrowOS 13.1 AppManager works.  The cache folder is created automatically.  The conclusion is that it must have been a bug in the previous version because even Scary Teacher 3D doesn't work on ArrowOS 13.0 and it does on 13.1.

**Comment by MuntashirAkon** [comment]
According to a discussion on our Telegram group ([1](https://t.me/c/1648630004/1906), [2](https://t.me/c/1648630004/1907)), the problem likely _stems from incorrectly formatting data after flashing the ROM_.

> We had this on later OnePlus devices. If you don't use the ROM default recovery to format, weird data access issues occur for external app data and cache

**Comment by MuntashirAkon** [comment]
Similar issue reported via E-Mail (TECNO, OnePlus):

<details>

<summary>Crash reports</summary>

```
java.lang.IllegalStateException: External directory unavailable.
    at io.github.muntashirakon.AppManager.ipc.RootServiceManager.startRootProcess(RootServiceManager.java:161)
    at io.github.muntashirakon.AppManager.ipc.RootServiceManager.createBindTask(RootServiceManager.java:285)
    at io.github.muntashirakon.AppManager.ipc.RootService.bindOrTask(RootService.java:146)
    at io.github.muntashirakon.AppManager.ipc.RootService.bind(RootService.java:110)
    at io.github.muntashirakon.AppManager.ipc.RootService.bind(RootService.java:125)
    at io.github.muntashirakon.AppManager.ipc.ServiceConnectionWrapper.lambda$startDaemon$0$io-github-muntashirakon-AppManager-ipc-ServiceConnectionWrapper(ServiceConnectionWrapper.java:117)
    at io.github.muntashirakon.AppManager.ipc.ServiceConnectionWrapper$$ExternalSyntheticLambda0.run(Unknown Source:4)
    at android.os.Handler.handleCallback(Handler.java:938)
    at android.os.Handler.dispatchMessage(Handler.java:99)
    at android.os.Looper.loopOnce(Looper.java:241)
    at android.os.Looper.loop(Looper.java:342)
    at android.app.ActivityThread.main(ActivityThread.java:8117)
    at java.lang.reflect.Method.invoke(Native Method)
    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:583)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1045)
Caused by: java.io.FileNotFoundException: No available shared storage found.
   at io.github.muntashirakon.AppManager.utils.FileUtils.getExternalCachePath(FileUtils.java:210)
   at io.github.muntashirakon.AppManager.ipc.RootServiceManager.startRootProcess(RootServiceManager.java:159)
   at io.github.muntashirakon.AppManager.ipc.RootServiceManager.createBindTask(RootServiceManager.java:285)
   at io.github.muntashirakon.AppManager.ipc.RootService.bindOrTask(RootService.java:146)
   at io.github.muntashirakon.AppManager.ipc.RootService.bind(RootService.java:110)
   at io.github.muntashirakon.AppManager.ipc.RootService.bind(RootService.java:125)
   at io.github.muntashirakon.AppManager.ipc.ServiceConnectionWrapper.lambda$startDaemon$0$io-github-muntashirakon-AppManager-ipc-ServiceConnectionWrapper(ServiceConnectionWrapper.java:117)
   at io.github.muntashirakon.AppManager.ipc.ServiceConnectionWrapper$$ExternalSyntheticLambda0.run(Unknown Source:4)
   at android.os.Handler.handleCallback(Handler.java:938)
   at android.os.Handler.dispatchMessage(Handler.java:99)
   at android.os.Looper.loopOnce(Looper.java:241)
   at android.os.Looper.loop(Looper.java:342)
   at android.app.ActivityThread.main(ActivityThread.java:8117)
   at java.lang.reflect.Method.invoke(Native Method)
   at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:583)
   at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1045)

Device Info:
App version: 3.1.0
App version code: 423
Android build version: 221120V757
Android release version: 12
Android SDK version: 31
Android build ID: CI6-H6917ABC-S-GL-221120V757
Device brand: TECNO
Device manufacturer: TECNO
Device name: TECNO-CI6
Device model: TECNO CI6
Device product name: CI6-GL
Device hardware name: mt6769
ABIs: [arm64-v8a, armeabi-v7a, armeabi]
ABIs (32bit): [armeabi-v7a, armeabi]
ABIs (64bit): [arm64-v8a]
System language: en-US
In-App Language: auto
Mode: auto
Inferred Mode: root 
```

```
java.lang.IllegalStateException: External directory unavailable.
    at io.github.muntashirakon.AppManager.ipc.RootServiceManager.startRootProcess(RootServiceManager.java:161)
    at io.github.muntashirakon.AppManager.ipc.RootServiceManager.createBindTask(RootServiceManager.java:285)
    at io.github.muntashirakon.AppManager.ipc.RootService.bindOrTask(RootService.java:146)
    at io.github.muntashirakon.AppManager.ipc.RootService.bind(RootService.java:110)
    at io.github.muntashirakon.AppManager.ipc.RootService.bind(RootService.java:125)
    at io.github.muntashirakon.AppManager.ipc.ServiceConnectionWrapper.lambda$startDaemon$0$io-github-muntashirakon-AppManager-ipc-ServiceConnectionWrapper(ServiceConnectionWrapper.java:117)
    at io.github.muntashirakon.AppManager.ipc.ServiceConnectionWrapper$$ExternalSyntheticLambda0.run(Unknown Source:4)
    at android.os.Handler.handleCallback(Handler.java:942)
    at android.os.Handler.dispatchMessage(Handler.java:99)
    at android.os.Looper.loopOnce(Looper.java:240)
    at android.os.Looper.loop(Looper.java:351)
    at android.app.ActivityThread.main(ActivityThread.java:8364)
    at java.lang.reflect.Method.invoke(Native Method)
    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:584)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1013)
Caused by: java.io.FileNotFoundException: No available shared storage found.
   at io.github.muntashirakon.AppManager.utils.FileUtils.getExternalCachePath(FileUtils.java:210)
   at io.github.muntashirakon.AppManager.ipc.RootServiceManager.startRootProcess(RootServiceManager.java:159)
   at io.github.muntashirakon.AppManager.ipc.RootServiceManager.createBindTask(RootServiceManager.java:285)
   at io.github.muntashirakon.AppManager.ipc.RootService.bindOrTask(RootService.java:146)
   at io.github.muntashirakon.AppManager.ipc.RootService.bind(RootService.java:110)
   at io.github.muntashirakon.AppManager.ipc.RootService.bind(RootService.java:125)
   at io.github.muntashirakon.AppManager.ipc.ServiceConnectionWrapper.lambda$startDaemon$0$io-github-muntashirakon-AppManager-ipc-ServiceConnectionWrapper(ServiceConnectionWrapper.java:117)
   at io.github.muntashirakon.AppManager.ipc.ServiceConnectionWrapper$$ExternalSyntheticLambda0.run(Unknown Source:4)
   at android.os.Handler.handleCallback(Handler.java:942)
   at android.os.Handler.dispatchMessage(Handler.java:99)
   at android.os.Looper.loopOnce(Looper.java:240)
   at android.os.Looper.loop(Looper.java:351)
   at android.app.ActivityThread.main(ActivityThread.java:8364)
   at java.lang.reflect.Method.invoke(Native Method)
   at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:584)
   at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1013)

Device Info:
App version: 3.1.0
App version code: 423
Android build version: R.ed8c2c_1_2
Android release version: 13
Android SDK version: 33
Android build ID: KB2001_11_F.63
Device brand: OnePlus
Device manufacturer: OnePlus
Device name: OnePlus8T
Device model: KB2001
Device product name: OnePlus8T_IND
Device hardware name: qcom
ABIs: [arm64-v8a, armeabi-v7a, armeabi]
ABIs (32bit): [armeabi-v7a, armeabi]
ABIs (64bit): [arm64-v8a]
System language: en-IN
In-App Language: auto
Mode: auto
Inferred Mode: root 
```

</details>

**Comment by iambeingtracked** [comment]
/storage/emulated/0/Android/data/io.github.muntashirakon.AppManager exists, but the same file ending with .debug doesn't

**Comment by MuntashirAkon** [comment]
Another user has reported that a faulty encryption caused the issue on their device.

**Comment by shuvashish76** [comment]
> All issues seem to be coming from Xioami phones with A12 and 13 (Yet another Motorola-like situation (#880)?)

1st time crashed for me, (Motorola device - stock ROM) while testing multiple features of AM simultaneously. Tried to reproduce it but able to make it crash twice so far. So I'm not sure exact steps, will report here if I successfully reproduce it.
<details><summary>Details</summary>
<p>

 ```
java.lang.IllegalStateException: External directory unavailable.
    at io.github.muntashirakon.AppManager.ipc.RootServiceManager.lambda$startRootProcess$0$io-github-muntashirakon-AppManager-ipc-RootServiceManager(RootServiceManager.java:183)
    at io.github.muntashirakon.AppManager.ipc.RootServiceManager$$ExternalSyntheticLambda5.run(Unknown Source:11)
    at io.github.muntashirakon.AppManager.ipc.RootService.lambda$asRunnable$0(RootService.java:201)
    at io.github.muntashirakon.AppManager.ipc.RootService$$ExternalSyntheticLambda0.run(Unknown Source:2)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
    at java.lang.Thread.run(Thread.java:920)
 Caused by: java.io.FileNotFoundException: No available shared storage found.
   at io.github.muntashirakon.AppManager.utils.FileUtils.getExternalCachePath(FileUtils.java:166)
   at io.github.muntashirakon.AppManager.ipc.RootServiceManager.lambda$startRootProcess$0$io-github-muntashirakon-AppManager-ipc-RootServiceManager(RootServiceManager.java:181)
   at io.github.muntashirakon.AppManager.ipc.RootServiceManager$$ExternalSyntheticLambda5.run(Unknown Source:11)
   at io.github.muntashirakon.AppManager.ipc.RootService.lambda$asRunnable$0(RootService.java:201)
   at io.github.muntashirakon.AppManager.ipc.RootService$$ExternalSyntheticLambda0.run(Unknown Source:2)
   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
   at java.lang.Thread.run(Thread.java:920)

Device Info:
App version: 4.0.0-alpha01-DEBUG
App version code: 434
Android build version: 8049b-fa5cf
Android release version: 12
Android SDK version: 31
Android build ID: S2RUBS32.51-15-9-7
Device brand: motorola
Device manufacturer: motorola
Device name: corfur
Device model: moto g71 5G
Device product name: corfur_g
Device hardware name: qcom
ABIs: [arm64-v8a, armeabi-v7a, armeabi]
ABIs (32bit): [armeabi-v7a, armeabi]
ABIs (64bit): [arm64-v8a]
System language: en-IN
In-App Language: auto
Mode: root
Inferred Mode: No root
```
</p>
</details>

**Comment by MuntashirAkon** [comment]
I think we should seriously think about the issue since a lot of users are facing it. If anybody facing the issue is willing to help me, please follow steps 2 and 3 of https://github.com/MuntashirAkon/AppManager/issues/948#issuecomment-1482344629 (in latest debug) and report here.

**Comment by 7ala** [comment]
switched from pe+ a13 to crdroid 10 a14 on my poco x3 nfc, never faced an issue on pe+, now the app crashes on 2.ii

**Comment by MuntashirAkon** [comment]
> switched from pe+ a13 to crdroid 10 a14 on my poco x3 nfc, never faced an issue on pe+, now the app crashes on 2.ii

Does it work with ADB?

**Comment by SK00RUPA** [comment]
> > switched from pe+ a13 to crdroid 10 a14 on my poco x3 nfc, never faced an issue on pe+, now the app crashes on 2.ii
> 
> Does it work with ADB?

A14 here too. Wireless debugging works (i.e. doesn't crash). Granting root permissions and enabling root mode in the settings result in a crash at the next app startup. At first glance logcat looks similiar to above mentioned issues.
[crash.txt](https://github.com/MuntashirAkon/AppManager/files/13801424/crash.txt)

**Comment by 7ala** [comment]
> > switched from pe+ a13 to crdroid 10 a14 on my poco x3 nfc, never faced an issue on pe+, now the app crashes on 2.ii
> 
> Does it work with ADB?

work like a charm, also noticed that the version 4.0.0-alpha02 doesn't crash and close, it get stuck on initializing screen on root mode

**Comment by unRekable** [comment]
> According to a discussion on our Telegram group ([1](https://t.me/c/1648630004/1906), [2](https://t.me/c/1648630004/1907)), the problem likely _stems from incorrectly formatting data after flashing the ROM_.
> 
> > We had this on later OnePlus devices. If you don't use the ROM default recovery to format, weird data access issues occur for external app data and cache

I cannot access the telegram messages, though I'm in all of the 3 channels. Is there any solution right now I can do myself without destroying my OS?

**Comment by iambeingtracked** [comment]
> > > switched from pe+ a13 to crdroid 10 a14 on my poco x3 nfc, never faced an issue on pe+, now the app crashes on 2.ii
> > 
> > 
> > Does it work with ADB?
> 
> work like a charm, also noticed that the version 4.0.0-alpha02 doesn't crash and close, it get stuck on initializing screen on root mode

+1, works without root gets stuck on the initialization screen with root

**Comment by MuntashirAkon** [comment]
At this point, I think we need to ensure that we're all on the same page. So, anyone facing troubles with root should report the following:
1. Does the latest AM Debug work in your device without root?
2. ... with root?
3. ... with ADB?
4. If 2 does not work, did you flash any interesting modules that I need to be aware of? Examples include gapps, *posed, etc.
5. SELinux context
6. Device encryption status

Thanks.

**Comment by Matrix-sudo** [comment]
Hey,

Which debug version should I test?

**Comment by Matrix-sudo** [comment]
Hello

So I found out the problem was that I had a bad /data partition after all. 
After a format everything is working.

**Comment by unRekable** [comment]
> Hello
> 
> So I found out the problem was that I had a bad /data partition after all. After a format everything is working.

Did you wipe all of your data? I don't wanna do that. 

@MuntashirAkon, tested with #2633
1. yes it does
2. nope, stucks and crashes
3. nope
4. not really, it never worked without anything flashed. i have lsposed but no modules there, nothing that should crash it
5. not sure, i think i saw it before in magisk but cannot say with 100%
6. also yes, its encrypted

**Comment by Matrix-sudo** [comment]
> > Hello
> > So I found out the problem was that I had a bad /data partition after all. After a format everything is working.
> 
> Did you wipe all of your data? I don't wanna do that.
> 
> @MuntashirAkon, tested with #2633
> 
> 1. yes it does
> 2. nope, stucks and crashes
> 3. nope
> 4. not really, it never worked without anything flashed. i have lsposed but no modules there, nothing that should crash it
> 5. not sure, i think i saw it before in magisk but cannot say with 100%
> 6. also yes, its encrypted

Sadly since you're encrypted, sounds like you have the same problem as me. Only formatting data will fix the permissions.
Use swift backup and copy your files.
Took me 2 hours and I was back online with my device and settings.

**Comment by unRekable** [comment]
> > > Hello
> > > So I found out the problem was that I had a bad /data partition after all. After a format everything is working.
> > 
> > 
> > Did you wipe all of your data? I don't wanna do that.
> > @MuntashirAkon, tested with #2633
> > 
> > 1. yes it does
> > 2. nope, stucks and crashes
> > 3. nope
> > 4. not really, it never worked without anything flashed. i have lsposed but no modules there, nothing that should crash it
> > 5. not sure, i think i saw it before in magisk but cannot say with 100%
> > 6. also yes, its encrypted
> 
> Sadly since you're encrypted, sounds like you have the same problem as me. Only formatting data will fix the permissions. Use swift backup and copy your files. Took me 2 hours and I was back online with my device and settings.

Hm, so Swift Backup can backup all app data etc? Are you encrypted again? If that could fix it, I might consider doing it. I just wonder what destroyed our /data partition, it's also weird we just have this with AM.

**Comment by Matrix-sudo** [comment]
> > > > Hello
> > > > So I found out the problem was that I had a bad /data partition after all. After a format everything is working.
> > > 
> > > 
> > > Did you wipe all of your data? I don't wanna do that.
> > > @MuntashirAkon, tested with #2633
> > > 
> > > 1. yes it does
> > > 2. nope, stucks and crashes
> > > 3. nope
> > > 4. not really, it never worked without anything flashed. i have lsposed but no modules there, nothing that should crash it
> > > 5. not sure, i think i saw it before in magisk but cannot say with 100%
> > > 6. also yes, its encrypted
> > 
> > 
> > Sadly since you're encrypted, sounds like you have the same problem as me. Only formatting data will fix the permissions. Use swift backup and copy your files. Took me 2 hours and I was back online with my device and settings.
> 
> Hm, so Swift Backup can backup all app data etc? Are you encrypted again? If that could fix it, I might consider doing it. I just wonder what destroyed our /data partition, it's also weird we just have this with AM.

Yes, swift backup can backup all app data. I stayed encrypted and that did the fix for me.
As of why? For me it was that when installing twrp on samsung phones you need to format data (bootloader unlock, flash recovery, flash boot repack so recovery doesn't get back to stock). The script I was running didn't format data on newer firmware - hence the bug appeared. After manual format - all is well

**Comment by MuntashirAkon** [comment]
> I just wonder what destroyed our /data partition, it's also weird we just have this with AM.

As I said earlier, there's a workaround, but I am more interested in finding out and fixing the actual reason (if possible) than applying it. This is not quite uncommon. Broken encryption being one of the possibilities was suggested by @androidacy, it was not my own discovery. However, it's still curious as it's affecting AM immediately after granting root. There's definitely something going on here, which may even be related to SELinux.

@unRekable: Can you also share some logs taken during the crash?

**Comment by droserasprout** [comment]
Hi! I have the same problem after updating to Android 14 (crDroid). Upgrade procedure included formatting data from recovery and flashing "legacy to retrofit dynamic" [ZIP](https://sourceforge.net/projects/crdroid/files/raphael/10.x/recovery/legacy%20to%20retrofit%20dynamic%20by%20%40raphael_alpha.zip/download) which changes the partition table with some script. May be related somehow. Ping me if you need more logs or other info.

```
java.lang.IllegalStateException: External directory unavailable.
    at io.github.muntashirakon.AppManager.ipc.RootServiceManager.startRootProcess(RootServiceManager.java:161)
    at io.github.muntashirakon.AppManager.ipc.RootServiceManager.createBindTask(RootServiceManager.java:285)
    at io.github.muntashirakon.AppManager.ipc.RootService.bindOrTask(RootService.java:146)
    at io.github.muntashirakon.AppManager.ipc.RootService.bind(RootService.java:110)
    at io.github.muntashirakon.AppManager.ipc.RootService.bind(RootService.java:125)
    at io.github.muntashirakon.AppManager.ipc.ServiceConnectionWrapper.lambda$startDaemon$0$io-github-muntashirakon-AppManager-ipc-ServiceConnectionWrapper(ServiceConnectionWrapper.java:117)
    at io.github.muntashirakon.AppManager.ipc.ServiceConnectionWrapper$$ExternalSyntheticLambda0.run(Unknown Source:4)
    at android.os.Handler.handleCallback(Handler.java:958)
    at android.os.Handler.dispatchMessage(Handler.java:99)
    at android.os.Looper.loopOnce(Looper.java:205)
    at android.os.Looper.loop(Looper.java:294)
    at android.app.ActivityThread.main(ActivityThread.java:8244)
    at java.lang.reflect.Method.invoke(Native Method)
    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:878)
 Caused by: java.io.FileNotFoundException: No available shared storage found.
   at io.github.muntashirakon.AppManager.utils.FileUtils.getExternalCachePath(FileUtils.java:210)
   at io.github.muntashirakon.AppManager.ipc.RootServiceManager.startRootProcess(RootServiceManager.java:159)
   at io.github.muntashirakon.AppManager.ipc.RootServiceManager.createBindTask(RootServiceManager.java:285)
   at io.github.muntashirakon.AppManager.ipc.RootService.bindOrTask(RootService.java:146)
   at io.github.muntashirakon.AppManager.ipc.RootService.bind(RootService.java:110)
   at io.github.muntashirakon.AppManager.ipc.RootService.bind(RootService.java:125)
   at io.github.muntashirakon.AppManager.ipc.ServiceConnectionWrapper.lambda$startDaemon$0$io-github-muntashirakon-AppManager-ipc-ServiceConnectionWrapper(ServiceConnectionWrapper.java:117)
   at io.github.muntashirakon.AppManager.ipc.ServiceConnectionWrapper$$ExternalSyntheticLambda0.run(Unknown Source:4)
   at android.os.Handler.handleCallback(Handler.java:958)
   at android.os.Handler.dispatchMessage(Handler.java:99)
   at android.os.Looper.loopOnce(Looper.java:205)
   at android.os.Looper.loop(Looper.java:294)
   at android.app.ActivityThread.main(ActivityThread.java:8244)
   at java.lang.reflect.Method.invoke(Native Method)
   at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
   at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:878)

Device Info:
App version: 3.1.4
App version code: 427
Android build version: 1706253636
Android release version: 14
Android SDK version: 34
Android build ID: UQ1A.240105.004
Device brand: Xiaomi
Device manufacturer: Xiaomi
Device name: raphael
Device model: Mi 9T Pro
Device product name: lineage_raphael
Device hardware name: qcom
ABIs: [arm64-v8a]
ABIs (32bit): []
ABIs (64bit): [arm64-v8a]
System language: en-US
In-App Language: auto
Mode: auto
Inferred Mode: root
```

**Comment by unRekable** [comment]
I hope this log helps you 

`java.lang.IllegalStateException: External directory unavailable.
    at io.github.muntashirakon.AppManager.ipc.RootServiceManager.lambda$startRootProcess$0$io-github-muntashirakon-AppManager-ipc-RootServiceManager(RootServiceManager.java:176)
    at io.github.muntashirakon.AppManager.ipc.RootServiceManager$$ExternalSyntheticLambda5.run(Unknown Source:11)
    at io.github.muntashirakon.AppManager.ipc.RootService.lambda$asRunnable$0(RootService.java:201)
    at io.github.muntashirakon.AppManager.ipc.RootService$$ExternalSyntheticLambda0.run(Unknown Source:2)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1137)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:637)
    at java.lang.Thread.run(Thread.java:1012)
 Caused by: java.io.FileNotFoundException: No available shared storage found.
   at io.github.muntashirakon.AppManager.utils.FileUtils.getExternalCachePath(FileUtils.java:192)
   at io.github.muntashirakon.AppManager.ipc.RootServiceManager.lambda$startRootProcess$0$io-github-muntashirakon-AppManager-ipc-RootServiceManager(RootServiceManager.java:174)
   at io.github.muntashirakon.AppManager.ipc.RootServiceManager$$ExternalSyntheticLambda5.run(Unknown Source:11)
   at io.github.muntashirakon.AppManager.ipc.RootService.lambda$asRunnable$0(RootService.java:201)
   at io.github.muntashirakon.AppManager.ipc.RootService$$ExternalSyntheticLambda0.run(Unknown Source:2)
   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1137)
   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:637)
   at java.lang.Thread.run(Thread.java:1012)

Device Info:
App version: 4.0.0-alpha02-DEBUG
App version code: 435
Android build version: V14.0.23.9.4.DEV
Android release version: 13
Android SDK version: 33
Android build ID: TKQ1.220829.002 release-keys
Device brand: Xiaomi
Device manufacturer: Xiaomi
Device name: star
Device model: M2102K1G
Device product name: star
Device hardware name: qcom
ABIs: [arm64-v8a, armeabi-v7a, armeabi]
ABIs (32bit): [armeabi-v7a, armeabi]
ABIs (64bit): [arm64-v8a]
System language: de-DE
In-App Language: auto
Mode: root
Inferred Mode: kein root`

**Comment by MuntashirAkon** [comment]
> I hope this log helps you
> 
> https://pastes.io/27ndbt72xu

I meant from logcat, not the crashlog.

**Comment by unRekable** [comment]
> > I hope this log helps you
> > https://pastes.io/27ndbt72xu
> 
> I meant from logcat, not the crashlog.

I hope that helps you out.
https://pastebin.com/trPiwruQ

I used adb logcat | findstr "AppManager"
If you want me to use specific settings, please let me know.

**Comment by unRekable** [comment]
> > I used adb logcat | findstr "AppManager"
> 
> Logcat cannot filter using package name, you can only filter by UID.

What's the command I should use then? And how do I get the IUD? Otherwise it's too much log

**Comment by quincyf467** [comment]
My device: Xiaomi Mi 9T Pro
device name: raphael
ROM: crDroid 10.2

Here is my logs from LogFox (another LogCat reader)

https://dpaste.com/4FPY3562M

**Comment by unRekable** [comment]
https://pastebin.com/Xwxa6RZ8 that are my logs from logfox.

**Comment by MuntashirAkon** [comment]
This is where it's failing: https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/core/java/android/app/ContextImpl.java;l=802;drc=2f1bbc2bb14dba88a0f033fb4ce32bbf20df25ed

Here are my remarks:
1. It says that `/storage/emulated/0/Android/data/io.github.muntashirakon.AppManager.debug/cache` folder does not exist
2. It tries to create this directory, set the necessary permissions but fails with an error code (e.g., -1) that is not present in the standard Unix header file
3. If the directory does not already exist, it's unable to create the directory in first place

Regardless of 3, this does look like a permission issue, and I hardly think there's anything I can do regarding this in the app as the problem is in the OS-level (probably caused by broken encryption/SELinux/root). So, I think the only alternative is the hack that I previously talked about and/or factory resetting the device.

**Comment by MuntashirAkon** [comment]
> Btw. I were able to run it with root mode, doing:
> 
> mkdir -p /storage/emulated/0/Android/data/io.github.muntashirakon.AppManager/cache chmod 777 /storage/emulated/0/Android/data/io.github.muntashirakon.AppManager/cache
> 
> Let's see if everything works!

Let me know how it goes. In addition, you could try the following after emptying the directory:

```sh
chmod 771 /storage/emulated/0/Android/data/io.github.muntashirakon.AppManager
chown $UID:1015 /storage/emulated/0/Android/data/io.github.muntashirakon.AppManager
restorecon /storage/emulated/0/Android/data/io.github.muntashirakon.AppManager
```

Replace `$UID` with the actual UID of App Manager.

**Comment by drogga** [comment]
I really wonder - why isn't the `cache`, especially `images` moved from `/storage/emulated/0/Android/data/io.github.muntashirakon.AppManager/` to where IMO they should've been in the first place: `/data/user/0/io.github.muntashirakon.AppManager`, along with their files content and problem solved (? probably ?), also in the case of the apps icons images cache, even tho it might create multiple instances of the exact same in order to make the listing of apps components and A.M. in general load fasted, I still don't think that every component's icon should be cached separately. I for example have 26 icons just for an app like BSPlayer, 24 the exact same and 2 same generic android ones.

**Comment by MuntashirAkon** [comment]
> I really wonder - why isn't the `cache`, especially `images` moved from `/storage/emulated/0/Android/data/io.github.muntashirakon.AppManager/` to where IMO they should've been in the first place: `/data/user/0/io.github.muntashirakon.AppManager`, along with their files content and problem solved (? probably ?), also in the case of the apps icons images cache, even tho it might create multiple instances of the exact same in order to make the listing of apps components and A.M. in general load fasted, I still don't think that every component's icon should be cached separately. I for example have 26 icons just for an app like BSPlayer, 24 the exact same and 2 same generic android ones.

The use of external cache directory is mandatory for privileged users.

**Comment by MuntashirAkon** [comment]
I'll try to add a workaround for it since the affected users seem to have a wide range of issues (e.g., encryption issue, kernel issue, bad ROM issue, and so on). This, I believe, has something to do with SELinux context because only this can explain all the problems as the issue occurs only after App Manager is granted root. The workaround, it appears, is simply creating the cache directory.

**Comment by MuntashirAkon** [comment]
Could anybody facing the issue post a crash log from 82de588c363216bef0bae72d779fb98f1d6a05f8? Thanks in advance.

**Comment by KinglyMinecon** [comment]
> Any updates?

I also have this issue
Am on lineage os galaxy a71
I have root and when granting access to app manager it crashes upon opening the app or trying to do anything with it

Where would I find logs?

**Comment by MuntashirAkon** [comment]
> Where would I find logs?

Crash reports are accessible through the notifications if you have them enabled.

**Comment by droserasprout** [comment]
>Any updates?
Nightlies work perfectly for my legacy-to-retrofit configuration above after your fixes.

**Comment by flerouwu** [comment]
I'm experiencing the same issues as OP with the latest stable build from F-Droid - v3.1.7 (430). Works fine if I don't grant it superuser access.

> Could anybody facing the issue post a crash log from [82de588](https://github.com/MuntashirAkon/AppManager/commit/82de588c363216bef0bae72d779fb98f1d6a05f8)? Thanks in advance.

The GitHub Actions artifacts from that commit have expired, but I built from source today (commit 906c498). Build was successful, and the app is no longer crashing after I grant it superuser access.

**Comment by MuntashirAkon** [comment]
> The GitHub Actions artifacts from that commit have expired, but I built from source today (commit [906c498](https://github.com/MuntashirAkon/AppManager/commit/906c4981ca8de26808f371b24c80e308e05ecd9c)). Build was successful, and the app is no longer crashing after I grant it superuser access.

You can also test with the latest debug builds.

**Comment by M2rsh** [comment]
For some reason after inserting an SD Card this issue fixed itself for me

**Comment by MuntashirAkon** [comment]
> For some reason after inserting an SD Card this issue fixed itself for me

That's because App Manager's internal working directory was switched from your internal sdcard to the external.

**Comment by singh9596** [comment]
> Btw. I were able to run it with root mode, doing:
> 
> mkdir -p /storage/emulated/0/Android/data/io.github.muntashirakon.AppManager/cache chmod 777 /storage/emulated/0/Android/data/io.github.muntashirakon.AppManager/cache
> 
> Let's see if everything works!

Your method works buddy. Thanks for providing the solution

**Comment by HanabishiRecca** [comment]
I'm facing this issue on Android 16 / experimental builds of LineageOS 23.
Tested on a completely fresh clean system with AppManager being the only user app installed.

I actually disabled both data encryption and SELinux - no difference. Workaround by manual dir creation works.
Other apps don't seem to have issues with creating their dirs inside `Android/data`.

**Comment by Dr34Dy** [comment]
I tried the methods mentioned, but none of them work on Lineage 23.2 (Android 16). It seems the system has blocked the options to create or modify folder or file permissions. I've tried using the console, Termux, Shell, and even the file manager itself (with root privileges), but it won't let me modify anything in the Android folder.

**Comment by MuntashirAkon** [comment]
> I tried the methods mentioned, but none of them work on Lineage 23.2 (Android 16). It seems the system has blocked the options to create or modify folder or file permissions. I've tried using the console, Termux, Shell, and even the file manager itself (with root privileges), but it won't let me modify anything in the Android folder.

I think it's a namespace issue. Even if you have root, it is running in local namespace rather than global namespace.

**Comment by HanabishiRecca** [comment]
I don't see any issues doing it in LineageOS via rooted debugging.

```
adb root
adb shell
cd /sdcard/Android/data/io.github.muntashirakon.AppManager/
mkdir cache
chown "$(stat -c '%u:%g' .)" cache
```

**Comment by MuntashirAkon** [comment]
> I don't see any issues doing it in LineageOS via rooted debugging.
> 
> ```
> adb root
> adb shell
> cd /sdcard/Android/data/io.github.muntashirakon.AppManager/
> mkdir cache
> chown "$(stat -c '%u:%g' .)" cache
> ```

ADB root shouldn't be affected as it uses global namespace by default.

**Comment by wefalltomorrow** [comment]
> Btw. I were able to run it with root mode, doing:
> 
> mkdir -p /storage/emulated/0/Android/data/io.github.muntashirakon.AppManager/cache chmod 777 /storage/emulated/0/Android/data/io.github.muntashirakon.AppManager/cache
> 
> Let's see if everything works!

This worked for me on Android 16, thank you!

**Comment by kicaBo** [comment]
Nothing of that worked for me on Lineage 23.2 Android 16, KSU root :(
Any other help please

EDIT: 

This module help me 

Universal-Storage-Fix-QPR2-V1_5-STABLE-RELEASE-1.zip

And 

adb root
adb shell
cd /sdcard/Android/data/io.github.muntashirakon.AppManager/
mkdir cache
chown "$(stat -c '%u:%g' .)" cache

**Comment by MuntashirAkon** [comment]
As I [mentioned above](https://github.com/MuntashirAkon/AppManager/issues/948#issuecomment-1450441734), the `vold` code I shared is indeed the primary issue here. When the cache directory does not exist, `ContextImpl` hands the job over to `vold` which is supposed to be creating that directory (and it does so correctly by issuing that `ioctl` syscall). But when App Manager is granted root, `ioctl` syscall breaks and `vold` fails to create the cache directory.

The fix is recreating the directory with correct permissions. I don't know if anything has changed since then, but the `vold` behavior seems to be remain unchanged, and thus, we'll continue to encounter this error for root solutions where `ioctl` (or its underlying behavior) is not handled correctly. It will not happen for any subdirectories though since those are not considered "special" directories and created without the need for `vold`.

Now that I have the answer, I'll implement the fix in the next release. This may not solve all the storage-related problems though.

**Comment by MuntashirAkon** [comment]
This is a test build containing the implementation: https://upload.disroot.org/r/dw1_1sGR#Uqw7LwZqN4DsIpbqU9p0ItAXgj0K+f2vNWhWngFTa9I=

Let me know if it works.

**Comment by Dr34Dy** [comment]
> This is a test build containing the implementation: https://upload.disroot.org/r/dw1_1sGR#Uqw7LwZqN4DsIpbqU9p0ItAXgj0K+f2vNWhWngFTa9I=
> 
> Let me know if it works.

Hello:

Thanks again for your contribution and for responding to all the requests.

I test ROMs on different devices and versions... The problem with the latest custom ROMs isn't just with your app, but with several apps. I think they changed the Android structure, which caused many apps to stop working correctly.

Personally, I opted to use versions that aren't the latest (but rather a pair of versions before to the last one released, since it's the latter that have the problems). I'm currently using CRDroid 12.5, which works well (the latest version, 12.9, doesn't). In Lineage, version 23.0 works well, while 23.2 has the aforementioned problems.

What you explained earlier about creating the folder and having the necessary permissions doesn't solve the problem in the latest versions of this ROM (LineageOS, CRdroid, eOS...). In fact, the folder is created and apparently has the necessary permissions. It seems that root access isn't implemented correctly in these versions, since several apps are experiencing these kinds of problems.

I don't recommend using the latest versions of these custom ROMs. Instead, use versions two or three years older than the latest. The changes in these ROMs aren't that significant; most are almost imperceptible.

Thanks for replying. Your app is excellent; it's one I always use on my devices.

Regards.

**Comment by MuntashirAkon** [comment]
I'm closing this in favor of #1967 since the other issue is related to the A16 QPR2+ root detection issue. This particular issue is considered fixed via f8d31264b317e4daae325f6a98724458f38dc3c5
